docs/current_docs/api/module-tests.mdx

```diff
@@ -0,0 +1,389 @@
+---
+slug: /api/module-tests
+---
+
+import Tabs from '@theme/Tabs';
+import TabItem from '@theme/TabItem';
+
+# Module Tests
+
+Like any other piece of software, Dagger Functions and modules should be thoroughly tested. This section documents proven patterns and best practices for effectively testing reusable modules.
+
+The following examples rely on a module called `greeter` that provides a function to greet a person:
+
+
+<Tabs groupId="language">
+<TabItem value="Go">
+
+```go file=./snippets/modules/testing/go/main.go
+```
+</TabItem>
+<TabItem value="Python">
+
+```python file=./snippets/modules/testing/python/src/greeter/main.py
+```
+
+</TabItem>
+<TabItem value="TypeScript">
+
+```typescript file=./snippets/modules/testing/typescript/src/main.ts
+```
+
+</TabItem>
+</Tabs>
+
+## Test module
+
+Well-written tests often provide the best documentation for your software, and this holds true for Dagger modules as well. It's considered a best practice to keep your tests close to the module's code so they can serve as both verification and reference. Additionally, tests that rely on the module's public API act as functional examples, clearly illustrating how to use the module.
+
+Following these principles leads to writing tests for your Dagger modules using Dagger modules themselves. In practice, this means creating a test module in the same directory as your main module and writing your tests as Dagger Functions, as shown below:
+
+<Tabs groupId="language">
+<TabItem value="Go">
+
+```bash
+mkdir tests
+cd tests
+dagger init --name tests --sdk go --source .
+```
+
+Then add the following to `main.go`:
+
+```go
+func (m *Tests) Hello(ctx context.Context) error {
```

Comment:
For all the inline code snippets, wdyt of importing the file instead of copy-pasting the code inline? It would probably be easier to maintain.

Reply:
Not all of the examples are functional and some I modified for brevity from the original code.

I think it would make more sense to get rid of the executable examples instead.

Reply:
My thoughts:
- I thought that the top-level module example was useful to give the user context for what is being tested later, I would keep it
- Instead of storing both the original code file and then using a modified inline example, it may be better to either only provide the inline code (and delete the full code file) or create a partial (and import it deleting the inline code) - the goal being to avoid duplication as much as possible